### PR TITLE
Use Welford's LLK for mean and var computation in Groupnorm

### PR DESCRIPTION
**Pull request by nsorabaTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/22019

### Problem description
🔧 The existing GroupNorm implementation used a two-pass algorithm for mean and variance computation, leading to increased memory usage and computational overhead. This approach required storing intermediate values and made multiple passes through the data, creating performance bottlenecks in normalization operations.

### What's changed
✨ Implemented Welford's online algorithm for GroupNorm mean and variance computation

**🚀 Single-pass computation**: Replaced two-pass algorithm with Welford's numerically stable online method
**💾 Reduced memory footprint**: Eliminated need to store intermediate results during computation  
**⚡ Enhanced performance**: Streamlined calculation process with lower computational overhead
**🎯 Improved numerical stability**: Better precision and reduced floating-point errors
**🔧 New LLK integration**: Added Welford's algorithm support in Low-Level Kernel (LLK) infrastructure
**📊 Mode selection**: Added configurable modes (legacy=0, welford=1, reciprocals=2) for backward compatibility

The changes include new compute kernels, updated dataflow patterns, and comprehensive test coverage to ensure correctness across different GroupNorm configurations.

Both L1 sharded and DRAM interleaved versions are supported.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/17788160774
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/17788158365
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [x] New/Existing tests provide coverage for changes
- [x] https://github.com/tenstorrent/tt-metal/actions/runs/17791073809

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_combine.h [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    WelfordStats<uint16_t> result;
+    result.mean = detail::float_to_bfloat16(overall.mean);
+    result.variance = detail::float_to_bfloat16(overall.variance);
+    result.count = overall.count;
+
+    DPRINT << "Combined mean(float): " << overall.mean << " (bf16): " << BF16(result.mean) << ENDL();
+    DPRINT << "Combined variance (float): " << overall.variance << " (bf16): " << BF16(result.variance) << ENDL();
+    DPRINT << "Combined count: " << overall.count << ENDL();
```
[the remark this answers is not in the document]
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_combine.h]: This is pending few more debug hours for https://github.com/tenstorrent/tt-metal/issues/28341
Will be resolved in next commit.

File: ttnn/ttnn/operations/normalization.py [hunk 2]
```diff
@@ -325,6 +325,40 @@ def create_group_norm_mask_impl(num_channel, num_groups, num_cores_across_channe
...
+        W: Width
+        num_groups: Number of groups
+        grid_x: Number of cores across x dimension
+        grid_y: Number of cores across y dimension
+
+    Returns:
+        Row major tensor with reciprocal values
+    """
+    import torch
+    import math
```
[the remark this answers is not in the document]
Comment (nsorabaTT) [reply ttnn/ttnn/operations/normalization.py]: We don't want to have any top level dependency in ttnn or torch, so this is by design.

File: tests/ttnn/unit_tests/operations/fused/test_group_norm.py [hunk 3]
```diff
@@ -200,14 +215,15 @@ def test_group_norm_with_block_sharded_v2_8x4_grid(device, N, C, H, W, num_group
...
+        use_welford=use_welford,
     )
 
     # output tensor
     output_tensor = ttnn.to_memory_config(output_tensor, ttnn.L1_MEMORY_CONFIG)
     output_tensor = ttnn.from_device(output_tensor)
     output_tensor = ttnn.to_torch(output_tensor)
 
-    assert_with_pcc(torch_output_tensor, output_tensor, 0.9997)
+    assert_with_pcc(torch_output_tensor, output_tensor, 0.9996)
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/fused/test_group_norm.py]: I don't see a comment in the description about the pcc being reduced. What's the reason?

File: tt_metal/include/compute_kernel_api/welford.h [hunk 4]
```diff
@@ -36,14 +38,18 @@ namespace ckernel {
...
+ *
  * @param current_row     The current row index (starting from 0). Should follow 0 <= current_row <= 32 and current_row
  * <= final_row. When current_row is 0, the previous mean and m2 are ignored.
  * @param final_row       The final row index. This index is not included in the update. Should follow current_row <=
  * final_row.A2D This dictates the total number of rows to update, starting from current_row.
  * @param num_skip_rows   Number of initial rows to skip in the update. Must be >= 0. Default is 0.
  *                        Setting this to a value greater than 0 skips the first num_skip_rows rows of the update.
  *                        Should follow 0 <= num_skip_rows <= 32.
- * @param reciprocal_lut_ptr       The reciprocal lookup table pointer for optimized speed
+ * @param reciprocal_lut       The optional reference to the reciprocal lookup table. If nullptr, the reciprocal will
```
Comment (bbradelTT) [inline tt_metal/include/compute_kernel_api/welford.h]: Isn't reciprocal_lut a reference?

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/groupnorm_op.cpp [hunk 5]
```diff
@@ -20,12 +20,13 @@ void GroupNorm::validate(
     const std::vector<Tensor>& input_tensors,
     const std::vector<std::optional<const Tensor>>& optional_input_tensors) const {
     TT_FATAL(
-        input_tensors.size() == 1 and optional_input_tensors.size() <= 4, "Must have between 1 to 5 input tensors");
+        input_tensors.size() == 1 and optional_input_tensors.size() <= 5, "Must have between 1 to 6 input tensors");
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/groupnorm_op.cpp]: It ouwld be good to print out the size().

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_combine.h [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    std::memcpy(&result, &float_bits, sizeof(float));
+    return result;
+}
+
+/**
+ * @brief Convert float to bfloat16 (stored as uint16_t).
+ * @param f Float value.
+ * @return bfloat16 representation as uint16_t.
+ */
+inline uint16_t float_to_bfloat16(float f) {
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_combine.h:37]: Can anything from https://github.com/tenstorrent/tt-metal/pull/28192 be utilized?

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_reader_mcast_sender_unary_gn.cpp [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+    // These are numbers in absolute terms, on a per batch, per group, per core basis without tiling
+    constexpr uint32_t num_channels_per_group = get_compile_time_arg_val(21);
+    constexpr uint32_t num_rows_per_group = get_compile_time_arg_val(22);
+
+    constexpr auto src0_args = TensorAccessorArgs<23>();
+    constexpr auto out_args = TensorAccessorArgs<src0_args.next_compile_time_args_offset()>();
+
+    constexpr uint32_t block_w_minus_one = block_w - 1;
+    constexpr uint32_t block_w_minus_two = block_w - 2;
+    constexpr uint32_t TILE_WIDTH = 32;
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_reader_mcast_sender_unary_gn.cpp]: Can we use existing defines?

**Review by bbradelTT — COMMENTED** [review]
Looks good. I'll wait for the DPRINT cleanup before approving.

File: ttnn/ttnn/operations/normalization.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nsorabaTT) [reply ttnn/ttnn/operations/normalization.py]: Removed math from local import as it is already a global import. Torch cannot be a global import, so that remain a local one.

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp [hunk 8]
```diff
@@ -1156,11 +1172,18 @@ operation::ProgramWithCallbacks groupnorm_multi_core(
         TT_FATAL(beta.value().layout() == Layout::ROW_MAJOR, "Beta tensor must have ROW_MAJOR layout");
     }
 
+    // Mode is 0 for legacy groupnorm, 1 for welford groupnorm, 2 for groupnorm with reciprocals
+    uint32_t groupnorm_mode = (reciprocals.has_value() ? 2 : use_welford ? 1 : 0);
```
[the remark this answers is not in the document]
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/multi_core/groupnorm_op_multi_core.cpp]: Added enum

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/writer_unary_gn_rm_gb.cpp [hunk 9]
```diff
@@ -37,9 +36,10 @@ void kernel_main() {
     constexpr uint32_t block_w = get_compile_time_arg_val(18);
     constexpr uint32_t block_hw = get_compile_time_arg_val(19);
 
-    constexpr uint32_t page_size = get_compile_time_arg_val(20);
+    constexpr uint32_t is_welford = get_compile_time_arg_val(20) > 0;
```
[the remark this answers is not in the document]
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/writer_unary_gn_rm_gb.cpp]: Changed to `use_welford`

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_combine.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_combine.h]: Removed all DPRINTs

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/groupnorm_op.cpp (hunk 5 above)
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/groupnorm_op.cpp]: Fixed in the latest change.

File: tests/ttnn/unit_tests/operations/fused/test_group_norm.py (hunk 3 above)
Comment (nsorabaTT) [reply tests/ttnn/unit_tests/operations/fused/test_group_norm.py]: This has been brought back to 0.9997 threshold when fp32 accumulation in dst registers in enabled.

In one case, (Height sharding), I need to reduce the PCC threshold by 0.0001 for welford's case. In this scenario,  don't see any noticeable difference in outputs. However, the tensor is small so in this case the PCC with welford's is coming out to be 0.99979 instead of 0.9998 or more.

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_combine.h (hunk 6 above)
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_combine.h:37]: I checked that PR. The changes are on the host side and not on the kernel side. It cannot be utilized now, but we can explore this in the future.

I created issue https://github.com/tenstorrent/tt-metal/issues/28668 to move this to `tt_metal/hw/inc/utils/bfloat16.h`

File: tt_metal/include/compute_kernel_api/welford.h [hunk 10]
```diff
@@ -26,26 +28,28 @@ namespace ckernel {
...
+ * @param num_skip_rows   Number of initial rows to skip in the update.
  *                        Setting this to a value greater than 0 skips the first num_skip_rows rows of the update.
- *                        Should follow 0 <= num_skip_rows <= 32.
- * @param reciprocal_lut_ptr       The reciprocal lookup table pointer for optimized speed
+ *                        Should follow 0 <= num_skip_rows <= TILE_HEIGHT(32).
+ * @param reciprocal_lut  The optional reference to the reciprocal lookup table. If nullptr, the reciprocal will
+ *                        be computed using float division.
  *
- * @note All 32 rows of the input tile are processed by this function.
+ * @note All TILE_WIDTH(32) columns of the input tile are processed by this function.
```
Comment (bbradelTT) [inline tt_metal/include/compute_kernel_api/welford.h]: nit: please add a space. Otherwise this looks like a function call.
```
TILEWIDTH (32)
```

**Review by bbradelTT — APPROVED** [review]

File: tt_metal/include/compute_kernel_api/welford.h (hunk 4 above)
Comment (bbradelTT) [reply tt_metal/include/compute_kernel_api/welford.h]: This comment probably still needs to be updated. There seems to be a disconnect between it saying nullptr and the variable being a reference.

Comment (nsorabaTT) [reply tt_metal/include/compute_kernel_api/welford.h]: I updated the comment to clarify what is passed in when reciprocal LUT is not used. Thanks for catching this.

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_reader_mcast_sender_unary_gn.cpp (hunk 7 above)
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/device/kernels/dataflow/welford_reader_mcast_sender_unary_gn.cpp]: Updated to use `TILE_WIDTH` from constants.hpp

File: tt_metal/include/compute_kernel_api/welford.h (hunk 10 above)
Comment (nsorabaTT) [reply tt_metal/include/compute_kernel_api/welford.h]: Done

**Comment by nsorabaTT** [comment]
Waiting on https://github.com/tenstorrent/tt-metal/pull/28674 to get merged before merging this in.

Pulled in LLK changes so All Post Commits can be run on this PR. This PR will be merged in with a rebase on top of main after #28674 is merged.

Otherwise everything else looks good to go.

**Comment by fvranicTT** [comment]
> Waiting on #28674 to get merged before merging this in.
> 
> Pulled in LLK changes so All Post Commits can be run on this PR. This PR will be merged in with a rebase on top of main after #28674 is merged.
> 
> Otherwise everything else looks good to go.

@nsorabaTT LLK change has been merged into metal 😉

**Review by amahmudTT — APPROVED** [review]

**Comment by nsorabaTT** [comment]
@tenstorrent/metalium-developers-ttnn-core Gentle reminder to review this PR. Thanks!

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/groupnorm.hpp [hunk 11]
```diff
@@ -21,14 +21,16 @@ struct ExecuteGroupNorm {
...
         const std::optional<MemoryConfig>& memory_config = std::nullopt,
         std::optional<ttnn::DataType> dtype = std::nullopt,
         std::optional<CoreGrid> core_grid = std::nullopt,
         std::optional<bool> inplace = std::nullopt,
         std::optional<ttnn::Layout> output_layout = std::nullopt,
         std::optional<int> num_out_blocks = std::nullopt,
         std::optional<DeviceComputeKernelConfig> compute_kernel_config = std::nullopt,
-        const std::optional<ttnn::Tensor>& negative_mask = std::nullopt);
+        const std::optional<ttnn::Tensor>& negative_mask = std::nullopt,
+        bool use_welford = false);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/normalization/groupnorm/groupnorm.hpp:33]: Please avoid bools like this in API. See contribution guide for details

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/groupnorm/groupnorm.hpp (hunk 11 above)
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/groupnorm.hpp:33]: I will a put in a follow up PR to address this by replacing it with an enum.

Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/groupnorm/groupnorm.hpp:33]: Added an issue to track this: https://github.com/tenstorrent/tt-metal/issues/28839
